logstash-core/src/main/java/org/logstash/plugins/pipeline/PipelineBus.java

```diff
@@ -31,30 +34,32 @@ public void sendEvents(final PipelineOutput sender,
                            final boolean ensureDelivery) {
         if (events.isEmpty()) return; // This can happen on pipeline shutdown or in some other situations
 
-        final ConcurrentHashMap<String, AddressState> addressesToInputs = outputsToAddressStates.get(sender);
-
-        addressesToInputs.forEach((address, addressState) -> {
-            final Stream<JrubyEventExtLibrary.RubyEvent> clones = events.stream().map(e -> e.rubyClone(RubyUtil.RUBY));
-
-            PipelineInput input = addressState.getInput(); // Save on calls to getInput since it's volatile
-            boolean sendWasSuccess = input != null && input.internalReceive(clones);
-
-            // Retry send if the initial one failed
-            while (ensureDelivery && !sendWasSuccess) {
-                // We need to refresh the input in case the mapping has updated between loops
-                String message = String.format("Attempted to send event to '%s' but that address was unavailable. " +
-                        "Maybe the destination pipeline is down or stopping? Will Retry.", address);
-                logger.warn(message);
-                input = addressState.getInput();
-                sendWasSuccess = input != null && input.internalReceive(clones);
-                try {
-                    Thread.sleep(1000);
-                } catch (InterruptedException e) {
-                    Thread.currentThread().interrupt();
-                    logger.error("Sleep unexpectedly interrupted in bus retry loop", e);
+        synchronized (sender) {
```

Comment:
Am I reading this correctly that only one output worker thread can send events this at a time? Will this effectively serialise the throughput when we have many workers?

If so, could this be resolved by instead using a `ReentrantReadWriteLock`'s `readLock` here where we read sender state and its `writeLock` in the other methods that modify the sender's state? This would allow many workers to send events concurrently while ensuring that the state isn't changed while it is being used.

Reply:
> Conceptually, I think that this changeset as-is will prevent the race condition(s) that cause stalling while a pipeline is being shut down. I am ok with merging as-is, and my remaining concerns are not blockers. I have a concern about the locking mechanism being used in PipelineBus#sendEvents effectively serialising the throughput of a pipeline with multiple workers.

Thanks, @yaauie. While some of the `OutputStrategyExt` implementations already serialize calls from worker threads to outputs, I agree that there is the possibility that for `OutputStrategyExt` implementations that do not serialize workers to outputs, there may be reduced throughput. I'd like to identify those specifically to ensure that the complexity of using `ReentrantReadWriteLocks` here was warranted. I will look into that in a follow-up PR.